lib/SILGen/SILGenExpr.cpp

```diff
@@ -3782,7 +3783,48 @@ visitMagicIdentifierLiteralExpr(MagicIdentifierLiteralExpr *E, SGFContext C) {
   llvm_unreachable("Unhandled MagicIdentifierLiteralExpr in switch.");
 }
 
-RValue RValueEmitter::visitCollectionExpr(CollectionExpr *E, SGFContext C) {
+RValue RValueEmitter::visitArrayExpr(ArrayExpr *E, SGFContext C) {
+  auto loc = SILLocation(E);
+  auto init = E->getInitializer();
+  auto *decl = dyn_cast<AbstractFunctionDecl>(init.getDecl());
+  Type elementType = E->getElementType();
+  ArrayRef<Identifier> argLabels = decl->getFullName().getArgumentNames();
+  CanType arrayTy = ArraySliceType::get(elementType)->getCanonicalType();
+  VarargsInfo varargsInfo =
+      emitBeginVarargs(SGF, loc, elementType->getCanonicalType(), arrayTy,
+                       E->getNumElements(), {});
+
+  for (unsigned index : range(E->getNumElements())) {
+    auto destAddr = varargsInfo.getBaseAddress();
+    if (index != 0) {
+      SILValue indexValue = SGF.B.createIntegerLiteral(
+          loc, SILType::getBuiltinWordType(SGF.getASTContext()), index);
+      destAddr = SGF.B.createIndexAddr(loc, destAddr, indexValue);
+    }
+    auto &destTL = varargsInfo.getBaseTypeLowering();
+    TemporaryInitialization init(destAddr, CleanupHandle::invalid());
+    ArgumentSource(E->getElements()[index])
+        .forwardInto(SGF, varargsInfo.getBaseAbstractionPattern(), &init,
```

Comment:
I thought about this more.

The emission of the expression here is the main source of potential evil with respect to making sure that we clean everything up. All other cleanups are either 1. not used, 2. consumed before we return the array, 3. the cleanup of the array itself (which we want to propagate to our caller).

The question is, what happens to cleanups generated from the element? Are we 100% sure that leaking those cleanups into the outside world is safe... I would argue that after we call the allocating initializer, we no longer require any of these cleanups to exist. Assuming that I am correct, it would suggest that the /conservative/ change here is to create an ArgumentScope and to push the result through the ArgumentScope. This will make sure that everything down stream of here has all formal/normal cleanups removed and that we push out just that singular array returned by the allocating initializer.

(IOW I would create an ArgumentScope at the beginning of the function and use popPreservingValue on that scope with the result of emitApplyAllocatingInitializer).